File: DSA/Assignment_5/div_rev.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node
{
    int data;
    struct Node *prev;
    struct Node *next;
}Node;

Node *createNode(int data)
{
    Node *newNode=(Node *)malloc(sizeof(Node));
    newNode->data=data;
    newNode->prev=NULL;
    newNode->next=NULL;
}

void insert_ele(Node* *head, int data)
{
    Node *newNode=createNode(data);
    if(*head==NULL)
    {
        *head=newNode;
        return;
    }
    Node *temp=*head;
    while(temp->next!=NULL)
    temp=temp->next;
    temp->next=newNode;
    newNode->prev=temp;
}

void print_ele(Node* *head)
{
    Node *temp=*head;
    while(temp)
    {
        printf("%d ",temp->data);
        temp=temp->next;
    }
    printf("\n");
}

void div(Node* *head, int len)
{
    int br;
    if(len%2==0)
    br=len/2;
    else
    br=(len/2)+1;

    Node *temp=*head;
    Node *head_1=NULL;
    Node *head_2=NULL;
    for(int i=1;i<=br;i++)
    {
        insert_ele(&head_1, temp->data);
        temp=temp->next;
    }
    for(int i=br+1;i<=len;i++)
    {
        insert_ele(&head_2, temp->data);
        temp=temp->next;
    }
    return head_1, head_2;
}

File: DSA/Assignment_5/div_rev.py
class Node:
    def __init__(self, data):
        self.data=data
        self.next=None
        self.prev=None

class DoublyLinkedList:
    def __init__(self):
        self.head=None

    def createNode(self, data):
        return Node(data)
    
    def insert_ele(self, data):
        newNode=self.createNode(data)
        if self.head is None:
            self.head=newNode
            return
        temp=self.head
        while temp.next:
            temp=temp.next
        temp.next=newNode
        newNode.prev=temp

    def print_list(self):
        if self.head is None:
            print("Empty List!")
            return
        temp=self.head
        while temp.next:
            print(temp.data,end=' ')
            temp=temp.next
        print()

    def div(self, len):
        if len%2==0:
            br=len/2
        else:
            br=len/2+1

        list1=DoublyLinkedList()
        list2=DoublyLinkedList()

        temp=self.head
        for _ in range(br):
            list1.insert_ele(temp.data)
            temp=temp.next
        for _ in range(br+1, len+1):
            list2.insert_ele(temp.data)
            temp=temp.next

        return list1, list2
    
    def rev(self):
        if self.head is None:
            print("Empty List!")
            return
        temp=self.head
        last=None
        while temp:
            swap=temp.next
            temp.next=temp.prev
            temp.prev=swap

            last=temp
            temp=temp.prev
        self.head=last

    def join(self, list2):
        if self.head is None:
            print("Empty List!")
            return
        temp=self.head
        while temp.next:
            temp=temp.next
        temp.next=list2.head
        if list2.head is not None:
            list2.head.prev=temp

def main():
    list=DoublyLinkedList()

    n=int(input("Enter number of nodes:"))
    for _ in range(n):
        ele=int(input("Enter element:"))
        list.insert_ele(ele)
    print("Original List:",end='')
    list.print_list()
    